### PR TITLE
Add ability to create and edit projects

### LINKED ISSUE #141: Add ability to create projects

### LINKED ISSUE #143: Add ability to edit projects

### PR DESCRIPTION
**Pull request by wdoug**
This PR closes #143, closes #141 

#### What does this PR do?

- Adds the ability to edit projects
- Adds the ability to create projects
- Adds higher order components for wrapping view, edit, and create pages

### CODE REVIEW

**Comment by wdoug** [comment]
I'm not super happy with the implementation of this right now. I'd like to refactor things so there is less boilerplate. I'm not sure exactly what path I want to go down for that though. Maybe adding [redux-form](https://redux-form.com) -- on one hand it seems a little silly to add redux just for forms, but on the other hand, it also opens up the door to using some redux ecosystem stuff, which could possibly make it easy to do things like persisting partially edited forms to local storage.

**Comment by wdoug** [comment]
Alright, I've refactored things, and while I think we can continue to make things better, they are way better than they started out, and I feel generally comfortable merging things as they stand now

File: src/sections/ProjectSection.js [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
-  }
-  const { name, description } = project;
+import EditableText from '../forms/EditableText';
+import EditableMarkdown from '../forms/EditableMarkdown';
 
+const ProjectSection = ({ formData, editing, onFormDataChange }) => {
+  const { name, description } = formData;
+  const commonProps = {
+    onChange: onFormDataChange,
+    editing
```
Comment (sungbae) [inline src/sections/ProjectSection.js:9]: Is this supposed to be a property? What does this do exactly?

File: src/sections/ProjectSection.js [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+  };
   return (
-    <div>
-      <h1>{name}</h1>
-      <p>{description}</p>
-    </div>
+    <React.Fragment>
+      <h1>
+        <EditableText value={name} label="Title" name="name" {...commonProps} />
```
Comment (sungbae) [inline src/sections/ProjectSection.js:14]: I guess you're injecting the elements into here (e.g. `{...commonProps}`)? But I still don't understand what `editing` is exactly...

File: src/utils/withCreatePage.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+          this.setState({ submitting: false });
+          alert(err);
+        }
+      };
+
+      _updateFormData = event => {
+        const { name, value } = event.target;
+        this.setState(state => ({
+          formData: {
+            ...state.formData,
```
Comment (sungbae) [inline src/utils/withCreatePage.js:73]: Is this how you can copy/paste the properties from `state.formData` here?

File: src/utils/withEditPage.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      };
+
+      componentWillReceiveProps(nextProps) {
+        if (nextProps.formData !== this.props.formData) {
+          this.setState({ formData: nextProps.formData });
+        }
+      }
+
+      render() {
+        if (!this.props.formData) {
```
Comment (sungbae) [inline src/utils/withEditPage.js:37]: Couldn't this always be false? If `this.props.formData` is initialized as `{}`.

File: src/utils/withViewPage.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * @returns (Component) => WrappedComponent
+ */
+export default function withViewPage(options) {
+  return WrappedComponent => {
+    class ViewPageWrapper extends React.PureComponent {
+      static displayName = `WithViewPage(${getDisplayName(WrappedComponent)})`;
+
+      render() {
+        const { formData } = this.props;
+        if (!formData) {
```
Comment (sungbae) [inline src/utils/withViewPage.js:21]: Same concern as the comment in `withEditPage.js`. If the `formData` is initialized already as `{}`, couldn't this always be false?

**Review by sungbae — CHANGES_REQUESTED** [review]
I think this is pretty solid. I just left a few questions!

File: src/sections/ProjectSection.js (hunk 1 above)
Comment (wdoug) [reply src/sections/ProjectSection.js:9]: It is shorthand for `editing: editing` in the object

File: src/utils/withCreatePage.js (hunk 3 above)
Comment (wdoug) [reply src/utils/withCreatePage.js:73]: It is a way of copying the keys and values from `formData` into this new object

File: src/utils/withEditPage.js (hunk 4 above)
Comment (wdoug) [reply src/utils/withEditPage.js:37]: It is passed in as a prop from above, which, as it is coded currently, initializes it as `undefined`

**Review by sungbae — APPROVED** [review]

**Comment by rufusasterisk** [comment]
Tested on the deploy/netlify link

##### Add new event
Adding a new event from /projects/create successfully created an event - link was visible on /projects
The event persisted after navigating away
The event body supports markdown

##### Edit existing event
Event saved changes made through edit process
Changes persisted after navigating away

**Pull request merged**